### PR TITLE
doc: TESTPLAN.md

### PR DESCRIPTION
**Pull request by justinmk3**
- start documenting test architecture / future plans
- inspired by https://github.com/aws/aws-toolkit-azure-devops/blob/master/docs/TESTPLAN.md

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docs/TESTPLAN.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    -   The `vscode` API is available, but a full VSCode instance is not started.
+    -   The Toolkit code is invoked as a library, not as an extension loaded in VSCode.
+    -   Call functions and create objects directly.
+    -   May mock state where needed, though this is discouraged in favor of "fake" data/objects/files.
+    -   May use the filesystem.
+    -   Main property is that [the test is fast](https://pycon-2012-notes.readthedocs.io/en/latest/fast_tests_slow_tests.html).
+-   Integration Tests: **slow** tests
+    -   Live in `src/integrationTest/`
+    -   Use a full instance of VSCode.
+    -   Global state is shared across tests, thus later tests can be polluted by earlier tests.
```
Comment (awschristou) [inline docs/TESTPLAN.md:31]: consider replacing "can be polluted by earlier tests" with "run the risk of being polluted by earlier tests"
(impact vs permission)

File: docs/TESTPLAN.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    -   Live in `src/test/`
+    -   The `vscode` API is available, but a full VSCode instance is not started.
+    -   The Toolkit code is invoked as a library, not as an extension loaded in VSCode.
+    -   Call functions and create objects directly.
+    -   May mock state where needed, though this is discouraged in favor of "fake" data/objects/files.
+    -   May use the filesystem.
+    -   Main property is that [the test is fast](https://pycon-2012-notes.readthedocs.io/en/latest/fast_tests_slow_tests.html).
+-   Integration Tests: **slow** tests
+    -   Live in `src/integrationTest/`
+    -   Use a full instance of VSCode.
```
Comment (awschristou) [inline docs/TESTPLAN.md:30]: ... "and have access to an activated instance of the extension"

File: docs/TESTPLAN.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+Ratio of unit to integ tests: 90% unit tests, 10% system/acceptance tests.
+
+## Test categories
+
+The test suite has two categories of tests:
+
+-   Unit Tests: **fast** tests
+    -   Live in `src/test/`
+    -   The `vscode` API is available, but a full VSCode instance is not started.
```
Comment (bryceitoc9) [inline docs/TESTPLAN.md:22]: We do start a full instance of VS Code, and some tests even utilize it (e.g. some of the webview-related ones, specifically tied to token replacement...should these be considered integration tests?).

File: docs/TESTPLAN.md (hunk 1 above)
Comment (bryceitoc9) [reply docs/TESTPLAN.md:31]: I don't think this is unique to our integration tests? May be worth a cleanup pass in the unit tests (although we typically figure it out quickly enough).

File: docs/TESTPLAN.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+VSCode documentation describes an [extension testing](https://code.visualstudio.com/api/working-with-extensions/testing-extension)
+approach. The Toolkit codebase uses that approach, except some
+modifications/workarounds in `src/test/testRunner.ts`.
+
+-   We use the [vscode-test](https://github.com/microsoft/vscode-test) package.
+    -   [Mocha](https://mochajs.org/) framework is used to write tests.
+-   New code requires new tests.
+
+## Testing Gaps
```
Comment (bryceitoc9) [inline docs/TESTPLAN.md:59]: Do we want/need a mention re: UI tests?

File: docs/TESTPLAN.md (hunk 3 above)
Comment (awschristou) [reply docs/TESTPLAN.md:22]: Our current unit tests are a mix of unit tests and mildly functional tests. In an ideal world we would go through our existing tests and split them along a more pure definition of unit tests, and move anything that leverages vs code experimental instances into a different group (possibly integration)

File: docs/TESTPLAN.md (hunk 1 above)
Comment (justinmk3) [reply docs/TESTPLAN.md:31]: cleanup pass ?

Comment (bryceitoc9) [reply docs/TESTPLAN.md:31]: We've definitely had unit tests write to global state before, and we only pick it up when it actually causes issues. I'm saying it's not unique to integration tests, but if we want it to be unique, we should go through our existing unit tests with a fine-toothed comb to see what we can find (and either correct or convert to integration tests)

File: docs/TESTPLAN.md (hunk 3 above)
Comment (justinmk3) [reply docs/TESTPLAN.md:22]: > In an ideal world we would go through our existing tests and split them along a more pure definition of unit tests, and move anything that leverages vs code experimental instances into a different group (possibly integration)

I am not sure that is ideal. Ideally all the tests run very fast, thus the distinction is meaningless. "Unit" is not technically the proper name for what we have, but I don't see why we would move anything.

**Pull request merged**